tools/export/gnuarmeclipse/__init__.py

```diff
@@ -0,0 +1,891 @@
+            shutil.rmtree(tmp_folder)
+
+        if ret_code == 0:
+            # Return Success
+            return 0
+
+        # Seems like something went wrong.
+        return -1
+
+    # -------------------------------------------------------------------------
+    # Process source files/folders exclusions.
+
+    def compute_exclusions(self):
+        """
+        With the project root as the only source folder known to CDT,
+        based on the list of source files, compute the folders to not
+        be included in the build.
+
+        The steps are:
+        - get the list of source folders, as dirname(source_file)
+        - compute the top folders (subfolders of the project folder)
+        - iterate all subfolders and add them to a tree, with all 
+        nodes markes as 'not used'
+        - iterate the source folders and mark them as 'used' in the
+        tree, including all intermediate nodes
+        - recurse the tree and collect all unused folders; descend
+        the hierarchy only for used nodes
+        """
+        source_folders = [self.filter_dot(s) for s in set(dirname(
+            src) for src in self.resources.c_sources + self.resources.cpp_sources + self.resources.s_sources)]
+        if '.' in source_folders:
+            source_folders.remove('.')
+
+        # print 'source folders'
+        # print source_folders
+
+        top_folders = [f for f in set(s.split(os.sep)[0]
+                                      for s in source_folders)]
+        # print 'top folders'
+        # print top_folders
+
+        self.source_tree = {}
+        for top_folder in top_folders:
+            for root, dirs, files in os.walk(top_folder):
+                if len(dirs) == 0:
+                    self.add_source_folder_to_tree(root)
+
+        for folder in source_folders:
+            self.add_source_folder_to_tree(folder, True)
+
+        # print
+        # print self.source_tree
+        # self.dump_paths(self.source_tree)
+
+        # print 'excludings'
+        self.excluded_folders = ['BUILD']
+        self.recurse_excludings(self.source_tree)
+
+        print 'Source folders: {0}, with {1} exclusions'.format(len(source_folders), len(self.excluded_folders))
+
+    def add_source_folder_to_tree(self, path, is_used=False):
+        """
+        Decompose a path in an array of folder names and create the tree.
+        On the second pass the nodes should be already there; mark them
+        as used.
+        """
+        # print path, is_used
+        a = path.split(os.sep)
+        n = self.source_tree
+        p = None
+        for s in a:
+            if s[0] == '.':
+                continue
+            if s not in n.keys():
+                nn = {}
+                nn['name'] = s
+                nn['children'] = {}
+                if p != None:
+                    nn['parent'] = p
+                n[s] = nn
+            n[s]['is_used'] = is_used
+            p = n[s]
+            n = n[s]['children']
```

Comment:
This is a bit hard to follow for two reasons:
1) one letter variable names
2) `n` and `nn` are used